In [1]:
import arff
import numpy as np
import pandas as pd

with open('titanic.arff', 'r', encoding='utf-8') as f:
    data = arff.load(f)

In [2]:
titanic = pd.DataFrame(data['data'], columns=[attr[0] for attr in data['attributes']])
titanic.head(20)

pclass survived                                               name  \
0      1.0        1                      Allen, Miss. Elisabeth Walton   
1      1.0        1                     Allison, Master. Hudson Trevor   
2      1.0        0                       Allison, Miss. Helen Loraine   
3      1.0        0               Allison, Mr. Hudson Joshua Creighton   
4      1.0        0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5      1.0        1                                Anderson, Mr. Harry   
6      1.0        1                  Andrews, Miss. Kornelia Theodosia   
7      1.0        0                             Andrews, Mr. Thomas Jr   
8      1.0        1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9      1.0        0                            Artagaveytia, Mr. Ramon   
10     1.0        0                             Astor, Col. John Jacob   
11     1.0        1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12     1.0        1                      Aubart, Mme. Leontine Pauline   
13     1.0        1                       Barber, Miss. Ellen 'Nellie'   
14     1.0        1               Barkworth, Mr. Algernon Henry Wilson   
15     1.0        0                                Baumann, Mr. John D   
16     1.0        0                           Baxter, Mr. Quigg Edmond   
17     1.0        1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18     1.0        1                              Bazzani, Miss. Albina   
19     1.0        0                               Beattie, Mr. Thomson   

       sex      age  sibsp  parch    ticket      fare    cabin embarked  boat  \
0   female  29.0000    0.0    0.0     24160  211.3375       B5        S     2   
1     male   0.9167    1.0    2.0    113781  151.5500  C22 C26        S    11   
2   female   2.0000    1.0    2.0    113781  151.5500  C22 C26        S  None   
3     male  30.0000    1.0    2.0    113781  151.5500  C22 C26        S  None   
4   female  25.0000    1.0    2.0    113781  151.5500  C22 C26        S  None   
5     male  48.0000    0.0    0.0     19952   26.5500      E12        S     3   
6   female  63.0000    1.0    0.0     13502   77.9583       D7        S    10   
7     male  39.0000    0.0    0.0    112050    0.0000      A36        S  None   
8   female  53.0000    2.0    0.0     11769   51.4792     C101        S     D   
9     male  71.0000    0.0    0.0  PC 17609   49.5042     None        C  None   
10    male  47.0000    1.0    0.0  PC 17757  227.5250  C62 C64        C  None   
11  female  18.0000    1.0    0.0  PC 17757  227.5250  C62 C64        C     4   
12  female  24.0000    0.0    0.0  PC 17477   69.3000      B35        C     9   
13  female  26.0000    0.0    0.0     19877   78.8500     None        S     6   
14    male  80.0000    0.0    0.0     27042   30.0000      A23        S     B   
15    male      NaN    0.0    0.0  PC 17318   25.9250     None        S  None   
16    male  24.0000    0.0    1.0  PC 17558  247.5208  B58 B60        C  None   
17  female  50.0000    0.0    1.0  PC 17558  247.5208  B58 B60        C     6   
18  female  32.0000    0.0    0.0     11813   76.2917      D15        C     8   
19    male  36.0000    0.0    0.0     13050   75.2417       C6        C     A   

     body                        home.dest  
0     NaN                     St Louis, MO  
1     NaN  Montreal, PQ / Chesterville, ON  
2     NaN  Montreal, PQ / Chesterville, ON  
3   135.0  Montreal, PQ / Chesterville, ON  
4     NaN  Montreal, PQ / Chesterville, ON  
5     NaN                     New York, NY  
6     NaN                       Hudson, NY  
7     NaN                      Belfast, NI  
8     NaN              Bayside, Queens, NY  
9    22.0              Montevideo, Uruguay  
10  124.0                     New York, NY  
11    NaN                     New York, NY  
12    NaN                    Paris, France  
13    NaN                             None  
14    NaN                    Hessle, Yorks  
15    NaN                  

### **Ilość poszczególnych etykiet zmiennych jakościowych**

In [3]:
print('Liczba etykiet zmiennej pclass:{}'.format(len(titanic['pclass'].unique())))
titanic.value_counts('pclass')

Liczba etykiet zmiennej pclass:3


pclass
3.0    709
1.0    323
2.0    277
Name: count, dtype: int64

In [4]:
print('Liczba etykiet zmiennej survived:{}'.format(len(titanic['survived'].unique())))
titanic.value_counts('survived')

Liczba etykiet zmiennej survived:2


survived
0    809
1    500
Name: count, dtype: int64

In [5]:
print('Liczba etykiet zmiennej sex:{}'.format(len(titanic['sex'].unique())))
titanic.value_counts('sex')

Liczba etykiet zmiennej sex:2


sex
male      843
female    466
Name: count, dtype: int64

In [6]:
print('Liczba etykiet zmiennej sibsp:{}'.format(len(titanic['sibsp'].unique())))
titanic.value_counts('sibsp') #liczba rodzenstwa lub malzonkow na pokladzie

Liczba etykiet zmiennej sibsp:7


sibsp
0.0    891
1.0    319
2.0     42
4.0     22
3.0     20
8.0      9
5.0      6
Name: count, dtype: int64

In [7]:
print('Liczba etykiet zmiennej parch:{}'.format(len(titanic['parch'].unique())))
titanic.value_counts('parch') #liczba rodzicow lub dzieci na pokladzie

Liczba etykiet zmiennej parch:8


parch
0.0    1002
1.0     170
2.0     113
3.0       8
4.0       6
5.0       6
6.0       2
9.0       2
Name: count, dtype: int64

In [8]:
print('Liczba etykiet zmiennej cabin:{}'.format(len(titanic['cabin'].unique())))
#titanic.value_counts('cabin')

Liczba etykiet zmiennej cabin:187


In [9]:
print('Liczba etykiet zmiennej embarked:{}'.format(len(titanic['embarked'].unique())))
titanic.value_counts('embarked')

Liczba etykiet zmiennej embarked:4


embarked
S    914
C    270
Q    123
Name: count, dtype: int64

In [10]:
print('Liczba etykiet zmiennej boat:{}'.format(len(titanic['boat'].unique())))
#titanic.value_counts('boat')

Liczba etykiet zmiennej boat:28


In [11]:
print('Liczba wszystkich pasażerów: {}'.format(len(titanic)))

Liczba wszystkich pasażerów: 1309


**Duża moc zbioru: sibsp, parch, cabin, boat:** Mogą być większym wyzwaniem dla modelu uczenia maszynowego, ale są też mniej znaczące. 

**Mała moc zbioru: pclass, survived, sex, embarked:** Łatwiejsze do przetworzenia przez model i mają większy wpływ na wynik zmiennej przewidywanej.

**Liczba wszystkich pasażerów: 1309 -** Jest to duża ilość, a nim więcej danych tym lepiej model może się nauczyć przewidywać.

#### **Unikalne etykiety zmiennej *cabin:***

In [12]:
print('Liczba unikalnych etykiet zmiennej cabin:{}'.format(len(titanic['cabin'].unique())))
print('Unikalne etykiety:')
print(titanic['cabin'].unique())
type(titanic['cabin'].unique())

Liczba unikalnych etykiet zmiennej cabin:187
Unikalne etykiety:
['B5' 'C22 C26' 'E12' 'D7' 'A36' 'C101' None 'C62 C64' 'B35' 'A23'
 'B58 B60' 'D15' 'C6' 'D35' 'C148' 'C97' 'B49' 'C99' 'C52' 'T' 'A31' 'C7'
 'C103' 'D22' 'E33' 'A21' 'B10' 'B4' 'E40' 'B38' 'E24' 'B51 B53 B55'
 'B96 B98' 'C46' 'E31' 'E8' 'B61' 'B77' 'A9' 'C89' 'A14' 'E58' 'E49' 'E52'
 'E45' 'B22' 'B26' 'C85' 'E17' 'B71' 'B20' 'A34' 'C86' 'A16' 'A20' 'A18'
 'C54' 'C45' 'D20' 'A29' 'C95' 'E25' 'C111' 'C23 C25 C27' 'E36' 'D34'
 'D40' 'B39' 'B41' 'B102' 'C123' 'E63' 'C130' 'B86' 'C92' 'A5' 'C51' 'B42'
 'C91' 'C125' 'D10 D12' 'B82 B84' 'E50' 'D33' 'C83' 'B94' 'D49' 'D45'
 'B69' 'B11' 'E46' 'C39' 'B18' 'D11' 'C93' 'B28' 'C49' 'B52 B54 B56' 'E60'
 'C132' 'B37' 'D21' 'D19' 'C124' 'D17' 'B101' 'D28' 'D6' 'D9' 'B80' 'C106'
 'B79' 'C47' 'D30' 'C90' 'E38' 'C78' 'C30' 'C118' 'D36' 'D48' 'D47' 'C105'
 'B36' 'B30' 'D43' 'B24' 'C2' 'C65' 'B73' 'C104' 'C110' 'C50' 'B3' 'A24'
 'A32' 'A11' 'A10' 'B57 B59 B63 B66' 'C28' 'E44' 'A26' 'A6' 'A7' 

numpy.ndarray

#### **Redukcja liczby cech:**

In [13]:
CabinReduced = titanic['cabin'].astype(str).str[0]
CabinReduced

0       B
1       C
2       C
3       C
4       C
       ..
1304    N
1305    N
1306    N
1307    N
1308    N
Name: cabin, Length: 1309, dtype: object

In [14]:
titanic['CabinReduced'] = CabinReduced
titanic[['cabin', 'CabinReduced']].head(20)

cabin CabinReduced
0        B5            B
1   C22 C26            C
2   C22 C26            C
3   C22 C26            C
4   C22 C26            C
5       E12            E
6        D7            D
7       A36            A
8      C101            C
9      None            N
10  C62 C64            C
11  C62 C64            C
12      B35            B
13     None            N
14      A23            A
15     None            N
16  B58 B60            B
17  B58 B60            B
18      D15            D
19       C6            C

In [15]:
print('Liczba unikalnych etykiet zmiennej cabin po redukcji:{}'.format(len(titanic['CabinReduced'].unique())))
print('Unikalne etykiety:')
print(titanic['CabinReduced'].unique())
type(titanic['CabinReduced'].unique())

Liczba unikalnych etykiet zmiennej cabin po redukcji:9
Unikalne etykiety:
['B' 'C' 'E' 'D' 'A' 'N' 'T' 'F' 'G']


numpy.ndarray

Dokonujemy redukcji liczby cech tej zmiennej, ponieważ kabiny oznaczone tą samą literą prawdopodobnie były do siebie podobne tzn. znajdowały się blisko siebie oraz miały podobny standard i wielkość. Szczegółowe informacje o konkretnym numerze kabiny w danej grupie nie są istotne, a nawet mogłyby negatywnie wpływać na naukę modelu. Redukcja pozwala nam zachować tę cechę i nadal uwzględniać ją przy pracy modelu, a jednocześnie wyciąga z niej najistotniejsze informacje. W tym zestawie danych taka operacja raczej nie powoduje negatywnych skutków, natomiast zawsze przed jej wykonaniem należy przemyśleć czy występuje ryzyko, że utracimy informacje, które miały znaczący wpływ na zmienną przewidywaną.

In [18]:
titanic.to_csv('titanic1904.csv', index=False)
